### PR TITLE
Support W&B sweeps

### LINKED ISSUE #382: Add support for W&B sweeps
### Problem / Goal

We would like to support W&B parameter sweeps from inside the pipeline.

### What does good look like?

- [ ] W&B sweeps supported

### Related context

_No response_

### PR DESCRIPTION
**Pull request by jemrobinson**
- Use Optuna to decide on per-trial hyperparameters
- Run each Optuna trial as a W&B job, connected to a sweep
- Use W&B UI to look at sweep performance
- Best Optuna trial tracked locally and printed at the end of each run

Closes #382

### CODE REVIEW

File: docs/src/user-guide/commands.md [hunk 1]
```diff
@@ -68,6 +68,24 @@ Checkpoints are saved to `${BASE_DIR}/training/wandb/run-<date>-<id>/checkpoints
     PYTORCH_ENABLE_MPS_FALLBACK=1 uv run imp train
     ```
 
+## `sweep initialise`
+
+```bash
+uv run imp sweep initialise --sweep-yaml example.sweep.yaml --config-name baseline/02_cnn_unet_cnn
+```
+
+Samples a batch of hyperparameter combinations with Optuna and writes configuration files to a local directory.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply docs/src/user-guide/commands.md]: Done in 67661c02

File: icenet_mp/cli/sweep.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    sweep_yaml: Annotated[
+        Path,
+        typer.Option("--sweep-yaml", help="Path to a sweep search-space YAML file"),
+    ],
+) -> None:
+    """Sample a batch of hyperparameter combinations and write a W&B sweep config.
+
+    Writes `trials.json` (the sampled overrides for each trial) and `wandb_sweep.yaml`
+    (a real W&B `method: grid` sweep over those trials) alongside the input file.
+    """
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/cli/sweep.py]: Done in 67661c02

File: icenet_mp/sweep/optuna_sampler.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                    }
+                }
+            }
+        )
+        return DictConfig(
+            OmegaConf.merge(base_config, parameter_overrides, wandb_overrides)
+        )
+
+    def initialise_study(self, model_cfg: DictConfig, sweep_id: str) -> None:
+        """Create an Optuna study and save the sampled trials to a JSON file."""
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/sweep/optuna_sampler.py]: Done in 67661c02

File: icenet_mp/cli/sweep.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    trainer = model.train(
+        checkpoint_dir=Path(checkpoint_dir).resolve() if checkpoint_dir else None,
+        multistage=multistage,
+    )
+
+    # Record the trial result and log the best trial
+    if not isinstance(ckpt := trainer.checkpoint_callback, ModelCheckpoint):
+        return
+    if ckpt.best_model_score is None:
+        return
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/cli/sweep.py:175]: Done in 167b6e0c

File: icenet_mp/sweep/optuna_sampler.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            "entity": self.entity,
+            "metric": self.metric,
+            "n_trials": self.n_trials,
+            "name": self.name,
+            "parameters": self.parameters,
+            "sampler": self.sampler_cls,
+            "seed": self.seed,
+        }
+        with (self.study_path / "optuna.yaml").open("w") as f_yaml:
+            yaml.dump(optuna_cfg, f_yaml, default_flow_style=False)
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/sweep/optuna_sampler.py]: Done in bc4c06c0

File: icenet_mp/cli/main.py [hunk 6]
```diff
@@ -31,6 +32,7 @@
 )
 app.add_typer(datasets_cli, name="datasets")
 app.add_typer(evaluation_cli)
+app.add_typer(sweep_cli, name="sweep")
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/cli/main.py:35]: Done in df046d64

**Comment by IFenton** [comment]
Reviewing this

**Review by IFenton — APPROVED** [review]
LGTM and it seems to run fine. A couple of thoughts on the documentation: 
 - It wasn't immediately obvious that the example.sweep file wouldn't work like the other yaml files, so you needed to give the full file path.
 - It also wasn't clear that the sweep folder would be saved in the base path. I guess you could remove the need for writing out the base path, though not sure if that would be more intuitive?

**Comment by jemrobinson** [comment]
> It wasn't immediately obvious that the example.sweep file wouldn't work like the other yaml files, so you needed to give the full file path.
> It also wasn't clear that the sweep folder would be saved in the base path. I guess you could remove the need for writing out the base path, though not sure if that would be more intuitive?

Done in d8bbb07

**Comment by IFenton** [comment]
LGTM

**Comment by jemrobinson** [comment]
> It wasn't immediately obvious that the example.sweep file wouldn't work like the other yaml files, so you needed to give the full file path.

@IFenton: would you prefer a different way of handling this? For example, specifying the location relative to the config directory, or including the sweep config as part of the main config file?

> It also wasn't clear that the sweep folder would be saved in the base path. I guess you could remove the need for writing out the base path, though not sure if that would be more intuitive?

Would you rather this was stored somewhere else? This is also where W&B runs are stored so I thought it made sense - perhaps it would be better to just require the **sweep ID** as an argument and let `imp` find the directory itself?

**Comment by IFenton** [comment]
> @IFenton: would you prefer a different way of handling this? For example, specifying the location relative to the config directory, or including the sweep config as part of the main config file?

I expected it to be with the other configs, as that seemed intuitive, so I can see the logic of specifying the location relative to the config directory (then presumably if people put it in the base level of that directory, they would just need to specify the config name which would seem easy). However, given it is now documented I don't think it is necessary to make that change. 

The alternative of adding it to the main config file seems unnecessary unless it is important for reproducing the run? Is the sweep config currently saved to WandB separately?

> Would you rather this was stored somewhere else? This is also where W&B runs are stored so I thought it made sense - perhaps it would be better to just require the sweep ID as an argument and let imp find the directory itself?

I think this is a logical place for it, I just wasn't expecting to find it there as the sweep config had a specified path, so I looked for it in that location. In terms of getting imp to find the directory from the sweep run, that could make sense, although if we do that, we should consider if we want to make the same change for the checkpoint (which we also currently specify a full path for).

**Comment by jemrobinson** [comment]
> I expected it to be with the other configs, as that seemed intuitive, so I can see the logic of specifying the location relative to the config directory (then presumably if people put it in the base level of that directory, they would just need to specify the config name which would seem easy). However, given it is now documented I don't think it is necessary to make that change.

OK

> The alternative of adding it to the main config file seems unnecessary unless it is important for reproducing the run? Is the sweep config currently saved to WandB separately?

Not as such, but the W&B sweep view shows you what each parameter was for each run

W B Chart 20_08_2026, 11_28_43

> I think this is a logical place for it, I just wasn't expecting to find it there as the sweep config had a specified path, so I looked for it in that location. In terms of getting imp to find the directory from the sweep run, that could make sense, although if we do that, we should consider if we want to make the same change for the checkpoint (which we also currently specify a full path for).

OK, let's leave this as is. The idea was that you could copy a checkpoint (or a sweep) to a new machine and carry on from there (or use someone else's on the same machine).
